src/collections/maple_tree/mod.rs

```diff
@@ -174,193 +175,327 @@ where
+        let leftShiftAmount: isize = distanceBetweenPivots as isize - pivotInsertCount as isize;
+
+        if distanceBetweenPivots != 0 {
+            if leftShiftAmount == -1 {
+                end = rangeEndPivotIndex as usize;
+            } else if includesMax {
+                end = R_CAPACITY - 1 - leftShiftAmount as usize;
+            } else {
+                end = rangeEndPivotIndex as usize - leftShiftAmount as usize;
+            }
+        } else {
+            end = rangeEndPivotIndex as usize + pivotInsertCount as usize;
+
+            if insertUpperPivot {
+                end -= 1;
+            }
+        }
+
+        end
+    }
+
+    pub fn make_space_for_insert(
+        &mut self,
+        distanceBetweenPivots: isize,
+        pivotInsertCount: usize,
+        rangeEndPivotIndex: isize,
+        includesMax: bool,
+        insertLowerPivot: bool,
+        insertUpperPivot: bool,
+    ) -> bool {
+        // start and end pivots of the range have pivots inbetween
+        if distanceBetweenPivots != 0 {
+            let leftShiftAmount: isize = distanceBetweenPivots as isize - pivotInsertCount as isize;
+
+            // pivotInsertCount can only ever be 2 so if distanceBetweenPivots is 1
             // leftShift amount could be -1 so we need to right shift by 1
             if leftShiftAmount == -1 {
-                if self.move_pivots_right(range.end as usize, 1) == false{
+                if self.move_pivots_right(rangeEndPivotIndex as usize, 1) == false {
                     return false;
                 }
-                end = range.end as usize; 
-            }
-            else if includesMax{
-                if self.move_pivots_left(R_CAPACITY-1, leftShiftAmount as usize) == false{
-                    return false; 
+            } else if includesMax {
+                if self.move_pivots_left(R_CAPACITY - 1, leftShiftAmount as usize) == false {
+                    return false;
                 }
-                end = R_CAPACITY-1-leftShiftAmount as usize;
-            }
-            else{
-                println!("leftShiftAmount = {}", leftShiftAmount);
-                if self.move_pivots_left(range.end as usize, leftShiftAmount as usize) == false{
+            } else {
+                if self.move_pivots_left(rangeEndPivotIndex as usize, leftShiftAmount as usize)
+                    == false
+                {
                     return false;
                 }
-                end = range.end as usize - leftShiftAmount as usize; 
             }
         }
-        else{
-            if includesMax && (shiftLowerPivot || shiftUpperPivot){
+        // pivots are adjacent
+        else {
+            if includesMax && (insertLowerPivot || insertUpperPivot) {
                 return false;
             }
-            if self.move_pivots_right(range.end as usize, pivotShiftCount as usize) == false{
+            if self.move_pivots_right(rangeEndPivotIndex as usize, pivotInsertCount as usize)
```

Comment:
Could do !self.move_ .... instead of == false?

Reply:
I will change this, I never really know which to use but I think I will just stick with !x instead of x == false